src/qgraphicsmozview_p.cpp

```diff
@@ -937,27 +927,61 @@ void QGraphicsMozViewPrivate::touchEvent(QTouchEvent* event)
+    QList<int> startIds(moveIds);
+
+    // Produce separate event for every pressed touch points
+    Q_FOREACH (int id, pressedIds) {
+        MultiTouchInput meventStart(MultiTouchInput::MULTITOUCH_START, timeStamp, TimeStamp(), 0);
+        startIds.append(id);
+        std::sort(startIds.begin(), startIds.end(), cmpInt);
+        Q_FOREACH (int startId, startIds) {
+            const QTouchEvent::TouchPoint& pt = event->touchPoints().at(idHash.value(startId));
+            mozilla::ScreenIntPoint nspt(pt.pos().x(), pt.pos().y());
+            meventStart.mTouches.AppendElement(SingleTouchData(pt.id(),
+                                                               nspt,
+                                                               mozilla::ScreenSize(1, 1),
+                                                               180.0f,
+                                                               pt.pressure()));
         }
-        Q_ASSERT(meventStart.mTouches.Length() > 0);
+
         ReceiveInputEvent(meventStart);
     }
-    if (meventMove.mTouches.Length()) {
-        if (meventStart.mTouches.Length()) {
-            meventMove.mTouches.AppendElements(meventStart.mTouches);
+
+    Q_FOREACH (int id, endIds) {
+        const QTouchEvent::TouchPoint& pt = event->touchPoints().at(idHash.value(id));
+        mozilla::ScreenIntPoint nspt(pt.pos().x(), pt.pos().y());
+        MultiTouchInput meventEnd(MultiTouchInput::MULTITOUCH_END, timeStamp, TimeStamp(), 0);
+        meventEnd.mTouches.AppendElement(SingleTouchData(pt.id(),
+                                                         nspt,
+                                                         mozilla::ScreenSize(1, 1),
+                                                         180.0f,
+                                                         pt.pressure()));
+        ReceiveInputEvent(meventEnd);
+    }
+
+    if (!moveIds.empty()) {
+        if (!pressedIds.empty()) {
+            moveIds.append(pressedIds);
+        }
+
+        // Sort touch lists by IDs just in case JS code identifies touches
+        // by their order rather than their IDs.
+        std::sort(moveIds.begin(), moveIds.end(), cmpInt);
+        MultiTouchInput meventMove(MultiTouchInput::MULTITOUCH_MOVE, timeStamp, TimeStamp(), 0);
+        Q_FOREACH (int id, moveIds) {
+            const QTouchEvent::TouchPoint& pt = event->touchPoints().at(idHash.value(id));
+            mozilla::ScreenIntPoint nspt(pt.pos().x(), pt.pos().y());
+            meventMove.mTouches.AppendElement(SingleTouchData(pt.id(),
+                                                              nspt,
+                                                              mozilla::ScreenSize(1, 1),
+                                                              180.0f,
+                                                              pt.pressure()));
         }
```

Comment:
Feels a bit odd that ends are sent before moves but it's just a logical thing if even that. It's only about fingers and which are reported first.